### PR TITLE
2427 internal users can transfer a facility or operation

### PR DESCRIPTION
**Pull request by Sepehr-Sobhani**
[ISSUE 1949](https://github.com/bcgov/cas-registration/issues/1949)
&&
[ISSUE 2427](https://github.com/bcgov/cas-registration/issues/2427)

🗒️ **NOTES:**

- Added `cas_analyst` as the new user role
- Excluded `all` from backend error util if error was a generic backend error (We don't want to see `All:<rest of the error>` in the UI)
- Removed unused jsonSchema
- Fixed `allowedRoles` prop on dashboard tile links
- Updated some fixtures
- We still need to add the cronjob (https://github.com/bcgov/cas-registration/issues/2040) to be able to transfer entities in the future

🧪 **TEST:**

- Log in as cas_analyst
- Click the Transfer link
- Click `Make a Transfer` button
- See transfer form

### CODE REVIEW

**Review by BCerki — COMMENTED** [review]
Made some comments, anything that has a `w` in it means I want to talk about it in the walkthrough

File: bc_obps/registration/schema/v2/transfer_event.py [hunk 1]
```diff
@@ -1,29 +1,18 @@
...
-    @staticmethod
-    def resolve_id(obj: Dict[str, Any]) -> UUID:
-        operation_id = obj.get('operation__id', None)
-        facility_id = obj.get('facilities__id', None)
-
-        record_id = operation_id if operation_id else facility_id
-        if not isinstance(record_id, UUID):
-            raise Exception('Missing valid UUID')
-        return record_id
+    id: UUID = Field(default_factory=uuid.uuid4)
```
[the remark this answers is not in the document]
Comment (BCerki) [reply bc_obps/registration/schema/v2/transfer_event.py:15]: From walkthrough, would end up with duplicate facility_id if transferred more than once

File: bc_obps/service/transfer_event_service.py [hunk 2]
```diff
@@ -30,3 +48,203 @@ def list_transfer_events(
...
+        """
+        Validates that there are no overlapping active transfer events for the given operation or facilities.
+        """
+        if operation_id:
+            # Check for overlapping transfer events with the operation
+            overlapping_operation = TransferEvent.objects.filter(
+                operation_id=operation_id,
+                status__in=[
+                    TransferEvent.Statuses.TO_BE_TRANSFERRED,
+                    TransferEvent.Statuses.COMPLETE,
```
[the remark this answers is not in the document]
Comment (BCerki) [reply bc_obps/service/transfer_event_service.py:65]: Still need to clarify what COMPLETE means--is it some kind of pending state? @andrea-williams

[the remark this answers is not in the document]
Comment (BCerki) [reply bc_obps/service/transfer_event_service.py:65]: Could use dates instead, seems possible for statuses and dates to get out of sync with each other

File: bciers/libs/actions/src/api/getOperationsByOperatorId.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { actionHandler } from "@bciers/actions";
+
+async function getOperationsByOperatorId(operatorId: string) {
```
[the remark this answers is not in the document]
Comment (BCerki) [reply bciers/libs/actions/src/api/getOperationsByOperatorId.ts:3]: Brianna can do this as part of operation status or separate ticket

**Comment by Sepehr-Sobhani** [comment]
> I'm making some notes here, not finished testing yet so I might rescind them as I get further along.
> 
> Is the transfers dashboard meant to have another tile? (If it is I suggest separate card and PR) image
> 

I'm not aware, so for now, I'm going to remove that sub-dashboard step because it doesn't make sense.

File: bc_obps/service/tests/test_facility_designated_operation_timeline_service.py [hunk 4]
```diff
@@ -106,3 +110,47 @@ def test_list_timeline_filter():
...
+
+class TestFacilityDesignatedOperationTimelineService:
+    @staticmethod
+    def test_get_current_timeline():
+        timeline_with_no_end_date = baker.make_recipe('utils.facility_designated_operation_timeline', end_date=None)
+        result_found = FacilityDesignatedOperationTimelineService.get_current_timeline(
+            timeline_with_no_end_date.operation_id, timeline_with_no_end_date.facility_id
+        )
+        assert result_found == timeline_with_no_end_date
+        timeline_with_end_date = baker.make_recipe(
```
[the remark this answers is not in the document]
Comment (Sepehr-Sobhani) [reply bc_obps/service/tests/test_facility_designated_operation_timeline_service.py:123]: I think so! I'm still not sure about other event types, but closure event could be an example.

File: bc_obps/service/tests/test_transfer_event_service.py [hunk 5]
```diff
@@ -20,3 +25,478 @@ def test_list_transfer_events():
...
+                operation_id=new_operation.id, facility_ids=[facility.id for facility in new_facilities]
+            )
+        except Exception as e:
+            pytest.fail(f"Unexpected exception raised: {e}")
+
+        # Scenario 2: Overlapping operation
+        with pytest.raises(Exception, match="An active transfer event already exists for the selected operation."):
+            TransferEventService.validate_no_overlapping_transfer_events(operation_id=operation.id)
+
+        # Scenario 3: Overlapping facilities
```
[the remark this answers is not in the document]
Comment (Sepehr-Sobhani) [reply bc_obps/service/tests/test_transfer_event_service.py:58]: that’s an excellent question—and also a terrifying one! Definitely worth discussing this 😵‍💫

File: bc_obps/service/transfer_event_service.py [hunk 6]
```diff
@@ -30,3 +48,203 @@ def list_transfer_events(
...
+
+        if facility_ids:
+            # Check for overlapping transfer events with the facilities
+            overlapping_facilities = TransferEvent.objects.filter(
+                facilities__id__in=facility_ids,
+                status__in=[
+                    TransferEvent.Statuses.TO_BE_TRANSFERRED,
+                    TransferEvent.Statuses.COMPLETE,
+                ],
+            ).distinct()
```
[the remark this answers is not in the document]
Comment (Sepehr-Sobhani) [reply bc_obps/service/transfer_event_service.py:79]: We still need this distinct. M2M fields can produce duplicate results (like the one we faced in the transfer data grid query).

File: bciers/apps/registration/app/components/transfers/TransferForm.tsx [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+    formState?.transfer_entity,
+  ]);
+
+  useEffect(() => {
+    handleFromOperationChange();
+  }, [formState?.from_operation]);
+
+  const submitHandler = async (e: IChangeEvent) => {
+    const updatedFormData = e.formData;
+    // we can't transfer facilities to the same operation
```
[the remark this answers is not in the document]
Comment (Sepehr-Sobhani) [reply bciers/apps/registration/app/components/transfers/TransferForm.tsx:208]: We allow the transfer of facilities between operations under the same operator, meaning from_operator and to_operator can be the same. However, we prevent users from selecting an operation if they are the same. 
I will also implement backend checks to enforce this validation.

File: bc_obps/service/tests/test_transfer_event_service.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Sepehr-Sobhani) [reply bc_obps/service/tests/test_transfer_event_service.py:58]: I created a ticket for this: https://github.com/bcgov/cas-registration/issues/2588
Feel free to refine it.

File: bciers/apps/registration/app/components/transfers/TransferForm.tsx [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+      setIsSubmitted(true);
+    }
+  };
+
+  return (
+    <>
+      {isSubmitted ? (
+        <TransferSuccess
+          fromOperatorId={formState.from_operator}
+          toOperatorId={formState.to_operator}
```
[the remark this answers is not in the document]
Comment (Sepehr-Sobhani) [reply bciers/apps/registration/app/components/transfers/TransferForm.tsx:243]: We are not storing the legal_name here, we just store IDs, that is why we need to pass the whole operators down.

**Review by BCerki — APPROVED** [review]
Great work on this!!

**Pull request merged**